google/cloud/firestore_v1/async_collection.py

```diff
@@ -0,0 +1,373 @@
+
+        Args:
+            document_data (dict): Property names and values to use for
+                creating the document.
+            document_id (Optional[str]): The document identifier within the
+                current collection. If not provided, an ID will be
+                automatically assigned by the server (the assigned ID will be
+                a random 20 character string composed of digits,
+                uppercase and lowercase letters).
+
+        Returns:
+            Tuple[:class:`google.protobuf.timestamp_pb2.Timestamp`, \
+                :class:`~google.cloud.firestore_v1.async_document.AsyncDocumentReference`]:
+                Pair of
+
+                * The ``update_time`` when the document was created/overwritten.
+                * A document reference for the created document.
+
+        Raises:
+            ~google.cloud.exceptions.Conflict: If ``document_id`` is provided
+                and the document already exists.
+        """
+        if document_id is None:
+            document_id = _auto_id()
+
+        document_ref = self.document(document_id)
+        write_result = await document_ref.create(document_data)
+        return write_result.update_time, document_ref
+
+    async def list_documents(self, page_size=None):
+        """List all subdocuments of the current collection.
+
+        Args:
+            page_size (Optional[int]]): The maximum number of documents
+            in each page of results from this request. Non-positive values
+            are ignored. Defaults to a sensible value set by the API.
+
+        Returns:
+            Sequence[:class:`~google.cloud.firestore_v1.collection.DocumentReference`]:
+                iterator of subdocuments of the current collection. If the
+                collection does not exist at the time of `snapshot`, the
+                iterator will be empty
+        """
+        parent, _ = self._parent_info()
+
+        iterator = self._client._firestore_api.list_documents(
+            parent,
+            self.id,
+            page_size=page_size,
+            show_missing=True,
+            metadata=self._client._rpc_metadata,
+        )
+        iterator.collection = self
+        iterator.item_to_value = _item_to_document_ref
+        return iterator
+
+    def select(self, field_paths):
+        """Create a "select" query with this collection as parent.
+
+        See
+        :meth:`~google.cloud.firestore_v1.async_query.AsyncQuery.select` for
+        more information on this method.
+
+        Args:
+            field_paths (Iterable[str, ...]): An iterable of field paths
+                (``.``-delimited list of field names) to use as a projection
+                of document fields in the query results.
+
+        Returns:
+            :class:`~google.cloud.firestore_v1.async_query.AsyncQuery`:
+            A "projected" query.
+        """
+        query = async_query.AsyncQuery(self)
```

Comment:
after looking at this, I think we may want to have a core class for collection with small modifications to allow async, since most of these things aren't async specific. Duplicated code seems likely to drift apart. wdyt @BenWhitehead

Reply:
SGTM! We can reduce bloat in the testers as well, should cut down line numbers significantly.